File: python/string/z_algo.py
def z_algo(s):
    n = len(s)
    z, l, r = [n]+[0]*(n-1), 0, 0
    for i in range(1, n):
        if i > r:
            l = r = i
            while r < n and s[r-l] == s[r]: r+= 1
            z[i] = r-l; r-= 1
            continue
        k = i-l
        if z[k] < r-i+1: z[i] = z[k]; continue
        l = i
        while r < n and s[r-l] == s[r]: r+= 1
        z[i] = r-l; r-= 1
    return z

File: cpp/string/z_algo.cpp
vector<int> z_algo(string& s){
	int n = s.size(), L=0, R=0;
	vector<int> z(n); z[0] = n;
	for(int i=1; i<n; i++){
		if(i > R){
			L = R = i;
			while(R < n && s[R-L] == s[R]) R++;
			z[i] = R-L; R--;
			continue;
		}
		int k = i - L;
		if(z[k] < R-i+1) z[i] = z[k];
		else{
			L = i;
			while(R < n && s[R-L] == s[R]) R++;
			z[i] = R-L; R--;
		}
	}
	return z;
}

template <typename T>
vector<int> z_algo(vector<T>& s){
	int n = s.size(), L=0, R=0;
	vector<int> z(n); z[0] = n;
	for(int i=1; i<n; i++){
		if(i > R){
			L = R = i;
			while(R < n && s[R-L] == s[R]) R++;
			z[i] = R-L; R--;
			continue;
		}
		int k = i - L;
		if(z[k] < R-i+1) z[i] = z[k];
		else{
			L = i;
			while(R < n && s[R-L] == s[R]) R++;
			z[i] = R-L; R--;
		}
	}
	return z;
}